core/src/main/scala/org/apache/spark/resource/ResourceUtils.scala

```diff
@@ -232,6 +274,37 @@ private[spark] object ResourceUtils extends Logging {
     resourceInfoMap
   }
 
+  /**
+   * This function is similar to getOrDiscoverallResources, except for it uses the ResourceProfile
+   * information instead of the application level configs.
+   *
+   * It first looks to see if resource were explicitly specified in the resources file
+   * (this would include specified address assignments and it only specified in certain
+   * cluster managers) and then it looks at the ResourceProfile to get
```

Comment:
What is the relation between resources file, ResourceProfile and discovery script?

All resources specified in ResourceProfile need to obtained via running discovery script?

Reply:
The ResourceProfile defines what resources the executor is supposed to have - the requirements.  4 cores, 1 GPU, 1 FPGA, etc.
The resource file is only used in standalone mode where the Worker is responsible for assigning resources to the executors.  The discovery script is used for any resources that aren't specified in the resource file.   You can actually mix them if you want - for instance in standalone mode lets say your workers know about GPUs, but not FPGA's.  The worker woudl start your executor passing a resources file with the GPU assignments, but then the executor would have to run the discovery script to find what FPGA's are available to it.

This PR isn't really changing how that is done, that is from the original accelerator aware scheduling work. This PR only changes where it gets the requirements from. Now it gets it from the ResourceProfile rather then the spark configs.

So to clarify a bit more Spark is relying on the cluster manager to assign containers with the resource you request. In the case of Standalone mode the Worker starts the executor assigning it resources via the resources file.  In the case of yarn and k8s, they just give you a container and don't tell you what resources are on it, so we use the discovery Script to find the addresses of those resources yarn/k8s gave you.

Reply:
Is it possible for yarn or k8s (or mesos ?) to give an executor which does not satisfy the resource constraint ? (in case of oversubscription of gpu's, etc ? We used to validate allocation for cpu/memory a while back ... probably does not do that anymore for cpu/mem)

Reply:
Not currently unless those resource managers are lying.  Right now for instance if you request GPUs from yarn and its not configured for it, it throws an exception.  I thought the same was true on k8s and we don't support on mesos. On the yarn side in the YarnAllocator we also set the resources when we get the matching requests so yarn would have to be telling us that the container it gave us matches with those resources.

If it did give us one and we accepted it, then the executor wouldn't start up if it didn't discovery enough resources that match the resource requirement.

Reply:
Sounds good, so this is our expectation that any RM should not give a container which is under-provisioned in terms of requested resources (over provisioning is fine I guess ? Like more memory or more GPU's than requested). Thanks for clarifying.

Reply:
that is correct, that should be considered a bug in the RM.  over provisioned is fine.